README.md

```diff
@@ -1,175 +1,404 @@
+
+An alternative approach, taken by [PG-MCP](https://github.com/stuzero/pg-mcp), is provide connection details via MCP tool calls at the time of use.
+This is more convenient for users who switch databases, and allows a single MCP server to simultaneously support multiple end-users.
+
+There must be a better approach than either of these.
+Both have security weaknesses—few MCP clients store the MCP server configuration securely (an exception is Goose), and credentials provided via MCP tools are passed through the LLM and stored in the chat history.
+Both also have usability issues in some scenarios.
+
+
+### Schema Information
+
+The purpose of the schema information tool is to provide the calling AI agent with the information it needs to generate correct and performant SQL.
+For example, suppose a user asks, "How many flights took off from San Francisco and landed in Paris during the past year?"
+The AI agent needs to find the table that stores the flights, the columns that store the origin and destinations, and perhaps a table that maps between airport codes and airport locations.
+
+
+*Why provide schema information tools when LLMs are generally capable of generating the SQL to retrieve this information from Postgres directly?*
+
+Our experience using Claude indicates that the calling LLM is very good at generating SQL to explore the Postgres schema by querying the [Postgres system catalog](https://www.postgresql.org/docs/current/catalogs.html) and the [information schema](https://www.postgresql.org/docs/current/information-schema.html) (an ANSI-standardized database metadata view).
+However, we do not know whether other LLMs do so as reliably and capably.
+
+*Would it be better to provide schema information using [MCP resources](https://modelcontextprotocol.io/docs/concepts/resources) rather than [MCP tools](https://modelcontextprotocol.io/docs/concepts/tools)?*
+
+The [Reference PostgreSQL MCP Server](https://github.com/modelcontextprotocol/servers/tree/main/src/postgres) uses resources to expose schema information rather than tools.
+Navigating a resources is similar to navigating a file system, so this approach is natural in many ways.
+However, resource support is not less widespread than tool support in the MCP client ecosystem (see [example clients](https://modelcontextprotocol.io/clients)).
+In addition, while the MCP standard says that resources can be accessed by either AI agents or end-user humans, some clients only support human navigation of the resource tree.
+
+
+### Protected SQL Execution
+
+*WIP*
+
+MCP clients have different mechanisms for protecting the SQL execution.
+
+SQL can be a powerful tool
+
+[Auto-run](https://docs.cursor.com/chat/tools#auto-run)
+
+
+
+## Postgres Pro Development
 
-## Development
+The instructions below are for developers who want to work on Postgres Pro, or users who prefer to install Postgres Pro from source.
 
 ### Local Development Setup
 
```

Comment:
We can add a "Getting Started" example walkthrough that builds on the "Usage Examples" section with more detailed steps.